packages/webhid/src/watcher.ts

```diff
@@ -0,0 +1,141 @@
+import { EventEmitter } from 'events'
+import { XKeys } from '@xkeys-lib/core'
+import { getOpenedXKeysPanels, setupXkeysPanel } from './methods'
+
+export type XKeysWatcherEvents = {
+	connected: (xkeysPanel: XKeys) => void
+	error: (err: unknown) => void
+}
+
+export declare interface XKeysWatcher {
+	// Note: This interface defines strong typings for any events that are emitted by the XKeysWatcher class.
+	on<U extends keyof XKeysWatcherEvents>(event: U, listener: XKeysWatcherEvents[U]): this
+	emit<U extends keyof XKeysWatcherEvents>(event: U, ...args: Parameters<XKeysWatcherEvents[U]>): boolean
+}
+
+const DEFAULT_POLLING_INTERVAL_MS = 1000
+
+/**
+ * Set up a watcher for newly connected X-keys panels.
+ * Note: It is highly recommended to set up a listener for the disconnected event on the X-keys panel, to clean up after a disconnected device.
+ */
+export class XKeysWatcher extends EventEmitter {
+	private pollingTimeout: number | undefined = undefined
+
+	public debug = false
+
+	private readonly seenHidDevices: Set<HIDDevice> = new Set()
+	/** A list of the devices we've called setupXkeysPanels for. */
+	private setupXKeysPanels: XKeys[] = []
+	private readonly hidDeviceToXKeysPanel: WeakMap<HIDDevice, XKeys> = new WeakMap()
+	private readonly hidDeviceToDisconnectedListener: WeakMap<HIDDevice, (...args: unknown[]) => void> = new WeakMap()
+
+	constructor(private readonly options?: XKeysWatcherOptions) {
+		super()
+		this.triggerUpdateConnectedDevices(true)
+		navigator.hid.addEventListener('disconnect', this.handleDisconnect)
+	}
+
+	/**
+	 * Stop the watcher.
+	 *
+	 * @param closeAllDevices Set to false in order to NOT close all devices. Use this if you only want to stop the watching. Defaults to true.
+	 */
+	public async stop(closeAllDevices = true): Promise<void> {
+		navigator.hid.removeEventListener('disconnect', this.handleDisconnect)
+
+		if (this.pollingTimeout !== undefined) {
+			clearTimeout(this.pollingTimeout)
+			this.pollingTimeout = undefined
+		}
+
+		if (closeAllDevices) {
+			// In order for an application to close gracefully,
+			// we need to close all devices that we've called setupXkeysPanel() on
+			await Promise.all(this.setupXKeysPanels.map((xKeys) => xKeys.close()))
+		}
+	}
+
+	private triggerUpdateConnectedDevices(immediate: boolean) {
+		this.pollingTimeout = (setTimeout as Window['setTimeout'])(
+			async () => {
+				try {
+					await this.updateConnectedDevices()
+				} catch (e) {
+					console.error(e)
+				}
+				this.triggerUpdateConnectedDevices(false)
+			},
+			immediate ? 0 : this.options?.pollingInterval ?? DEFAULT_POLLING_INTERVAL_MS
+		)
```

Comment:
You must not pass an async funtion to setTimeout, as a thrown error will result in an uncaught Error.

```suggestion
		this.pollingTimeout = (setTimeout as Window['setTimeout'])(
			() => {
				this.updateConnectedDevices()
					.then(() => {
						this.triggerUpdateConnectedDevices(false)
					})
					.catch((e) => {
						this.emit('error', e)
					})
			},
			immediate ? 0 : this.options?.pollingInterval ?? DEFAULT_POLLING_INTERVAL_MS
		)
```